src/Microsoft.IdentityModel.JsonWebTokens/JwtTokenUtilities.cs

```diff
@@ -165,23 +171,59 @@ internal static string DecompressToken(byte[] tokenBytes, string algorithm)
                     continue;
                 }
 
-                if (!cryptoProviderFactory.IsSupportedAlgorithm(decryptionParameters.Enc, key))
-                {
-                    algorithmNotSupportedByCryptoProvider = true;
-                    LogHelper.LogWarning(TokenLogMessages.IDX10611, LogHelper.MarkAsNonPII(decryptionParameters.Enc), key);
-                    continue;
-                }
-
                 try
                 {
-                    Validators.ValidateAlgorithm(decryptionParameters.Enc, key, jwtToken, validationParameters);
-                    decryptedTokenBytes = DecryptToken(cryptoProviderFactory, key, decryptionParameters);
-                    X509SecurityKey x509Key = key as X509SecurityKey;
-                    if (x509Key != null)
-                        LogHelper.LogInformation(TokenLogMessages.IDX10903, x509Key.Certificate?.Thumbprint);
-
-                    decryptionSucceeded = true;
-                    break;
+                    // The JsonWebTokenHandler will set the JsonWebToken and those values will be used.
+                    // The JwtSecurityTokenHandler will calculate values and set the values on DecrytionParameters.
+
+                    // JsonWebToken from JsonWebTokenHandler
+                    if (securityToken is JsonWebToken jsonWebToken)
+                    {
+                        if (!cryptoProviderFactory.IsSupportedAlgorithm(jsonWebToken.Enc, key))
+                        {
+                            algorithmNotSupportedByCryptoProvider = true;
+                            LogHelper.LogWarning(TokenLogMessages.IDX10611, LogHelper.MarkAsNonPII(decryptionParameters.Enc), key);
+                            continue;
+                        }
+
+                        Validators.ValidateAlgorithm(jsonWebToken.Enc, key, securityToken, validationParameters);
+                        decryptedTokenBytes = DecryptToken(
+                            cryptoProviderFactory,
+                            key,
+                            jsonWebToken.Enc,
+                            jsonWebToken.CipherTextBytes,
+                            jsonWebToken.HeaderAsciiBytes,
+                            jsonWebToken.InitializationVectorBytes,
+                            jsonWebToken.AuthenticationTagBytes);
+
+                        zipAlgorithm = jsonWebToken.Zip;
+                        decryptionSucceeded = true;
+                        break;
+                    }
+                    // JwtSecurityToken from JwtSecurityTokenHandler
```

Comment:
This is same as the JwtSecurityToken block. Can the properties of decryptionParameters be set in the calling method, JsonWebTokenhandler.DecrytTOken() line 765 so the code does not need to be duplicated?